include/ginkgo/core/config/property_tree.hpp

```diff
@@ -0,0 +1,169 @@
+
+#ifndef GKO_PUBLIC_CORE_CONFIG_PROPERTY_TREE_HPP_
+#define GKO_PUBLIC_CORE_CONFIG_PROPERTY_TREE_HPP_
+
+
+#include <cassert>
+#include <exception>
+#include <map>
+#include <stdexcept>
+#include <string>
+#include <type_traits>
+#include <vector>
+
+
+#include <ginkgo/core/config/data.hpp>
+
+
+namespace gko {
+namespace config {
+
+
+/**
+ * pnode is to describe the property tree
+ */
+class pnode {
+public:
+    using key_type = std::string;
+    using data_type = data;
+
+    enum status_t { empty, array, object, list };
+
+    pnode() : status_(status_t::empty) {}
+
+    pnode(const data_type& d) : status_(status_t::object), data_(d) {}
+
+    pnode(const std::vector<pnode>& array)
+        : status_(status_t::array), array_(array)
+    {}
+
+    pnode(const std::map<key_type, pnode>& list)
+        : status_(status_t::list), list_(list)
+    {}
+
+    // Get the content of node
+    data_type& get_data()
+    {
+        assert(status_ == status_t::object);
+        return data_;
+    }
+
+    const data_type& get_data() const
+    {
+        assert(status_ == status_t::object);
+        return data_;
+    }
+
+    std::vector<pnode>& get_array()
+    {
+        assert(status_ == status_t::array || status_ == status_t::empty);
+        status_ = status_t::array;
+        return array_;
+    }
+
+    const std::vector<pnode>& get_array() const
+    {
+        assert(status_ == status_t::array);
+        return array_;
+    }
+
+    std::map<key_type, pnode>& get_list()
+    {
+        assert(status_ == status_t::list || status_ == status_t::empty);
+        status_ = status_t::list;
+        return list_;
+    }
+
+    const std::map<key_type, pnode>& get_list() const
+    {
+        assert(status_ == status_t::list);
+        return list_;
+    }
+
+    // Get the data of node's content
+    template <typename T>
+    T get_data() const
+    {
+        assert(status_ == status_t::object);
+        return gko::config::get<T>(data_);
+    }
+
+    pnode& at(const std::string& path)
+    {
+        assert(status_ == status_t::list);
+        return list_.at(path);
+    }
+
+    const pnode& at(const std::string& path) const
+    {
+        assert(status_ == status_t::list);
+        return list_.at(path);
+    }
+
+    pnode& at(int i)
+    {
+        assert(status_ == status_t::array);
+        return array_.at(i);
+    }
+
+    const pnode& at(int i) const
+    {
+        assert(status_ == status_t::array);
+        return array_.at(i);
+    }
+
+    // Check the status
+    bool is(status_t s) const { return this->get_status() == s; }
+
+    bool contains(std::string key) const
+    {
+        assert(status_ == status_t::list);
+        auto it = list_.find(key);
+        return (it != list_.end());
+    }
+
+    status_t get_status() const { return status_; }
+
+private:
+    std::vector<pnode> array_;        // for array
+    std::map<key_type, pnode> list_;  // for list
+    data_type data_;                  // for value
```

Comment:
why do we have a separate implementation for primitive types and composite types? We can pull the implementation of `data_type` into this file directly.

Reply:
no, variant can not do the recursive variant

Reply:
Yes, it can also be handled by flattening design.
With the current design, (vector, map, value) describe the structure clearly. then all actual data will be data_type